Keyboard focusing/navigation support for the UWP platform.

reseul opened the pull request:
Existing React Native for UWP uses Border/Canvas based implementation for RCTView, with no support for keyboard input of any kind.
An updated version (covered by a concurrent PR) contains a helper FocusableWindows component that maps to a UserControl/Canvas based native view, component that can be used to add keyboard input/focusing capability to other components.

Main Changes:
1. "windows" flavors of View, Button, Link, TextInput, and ScrollViewer. We keep "Text" as is, with a "focus" method used strictly for narrator purposes.
2. The existing Web FocusManager is now refactored into:
	- Common implementation (in common/utils)
	- Web and Desktop flavors through derived classes.
   Methods marked as "xxxx /* static */" are logically static, yet Typescript is not nice with us on that regard..
3. A native-desktop folder containing what would be common between uwp and the macos platforms: App, RootView, UserInterface, etc.
Some more refined refactoring may happen when macos platform is merged in.

mshoho commented on src/common/utils/FocusManager.ts:
```diff
@@ -0,0 +1,387 @@
+    protected abstract /* static */ removeFocusListenerFromComponent(component: React.Component<any, any>, onFocus: () => void): void;
+    protected abstract /* static */ focusComponent(component: React.Component<any, any>): boolean;
+
+    // protected abstract /* static */ focusFirst(last?: boolean): void;
```
Why is static commented and why is focusFirst commented?
If there is a reason, I guess we need to remove the commented code completely.

reseul replied:
Yeah, right, a leftover. Each flavor has its own focusFirst called from the corresponding resetFocus, etc.

---
In reply to: [159244518](https://github.com/Microsoft/reactxp/pull/404#discussion_r159244518) [](ancestors = 159244518)

mshoho commented on src/native-common/TextInput.tsx:
```diff
@@ -48,51 +48,60 @@ export class TextInput extends React.Component<Types.TextInputProps, TextInputSt
+        const blurOnSubmit = this.props.blurOnSubmit || !this.props.multiline;
+
+        const internalProps: RN.TextInputProps = {
+            ref: 'nativeTextInput',
```
Refs by string are deprecated. And it makes sense to upgrade the previously existing refs by string to callbacks in some files above too, since you're touching the files anyway.

reseul replied:
Ok. I knew this will come..

---
In reply to: [159245310](https://github.com/Microsoft/reactxp/pull/404#discussion_r159245310) [](ancestors = 159245310)

rigdern commented on src/native-common/Button.tsx:
```diff
@@ -85,7 +85,7 @@ export class Button extends React.Component<Types.ButtonProps, {}> {
     private _isMounted = false;
     private _hideTimeout: number|undefined;
-    private _buttonElement: RN.Animated.View|undefined;
+    private _buttonElement: RN.Animated.View|null = null;
```
>null [](start = 45, length = 4)

Why do we switch this from `undefined` to `null`?

rigdern commented on src/native-common/utils/EventHelpers.ts:
```diff
@@ -0,0 +1,144 @@
+
+                    case 'Delete':
+                        keyCode = 8;
+                        break;
```
Should we move this to be with "Backspace" above which has the same key code?

reseul replied:
Ok, will do

---
In reply to: [159335721](https://github.com/Microsoft/reactxp/pull/404#discussion_r159335721) [](ancestors = 159335721)

rigdern commented on src/native-common/Button.tsx:
```diff
@@ -117,27 +128,25 @@ export class Button extends React.Component<Types.ButtonProps, {}> {
+        let internalProps: RN.ViewProps = {
+            ref: this._onButtonRef,
+            style: Styles.combine([_styles.defaultButton, this.props.style, opacityStyle,
+                            this.props.disabled && _styles.disabled]),
```
>this.props.disabled && _styles.disabled]), [](start = 28, length = 42)

Is this line indented too much? Should be only be indented 4 spaces relative to the previous line?

reseul replied:
ok. Was following the "keep close to the first param position" school of thought.

---
In reply to: [159335965](https://github.com/Microsoft/reactxp/pull/404#discussion_r159335965) [](ancestors = 159335965)

rigdern commented on src/windows/Button.tsx:
```diff
@@ -0,0 +1,211 @@
+});
+
+// Simple check for the presence of the updated React Native for Windows
+const IsUpdatedReactNativeForWindows = (RNW.FocusableWindows !== undefined);
```
>IsUpdatedReactNativeForWindows [](start = 6, length = 30)

Consider calling this "HasFocusableWindows" or something similar. I think it's better to think in terms of the APIs the platform supports rather than thinking in terms of versions.

reseul replied:
ok

---
In reply to: [159336672](https://github.com/Microsoft/reactxp/pull/404#discussion_r159336672) [](ancestors = 159336672)

rigdern commented on src/windows/Button.tsx:
```diff
@@ -0,0 +1,211 @@
+                isTabStop: windowsTabFocusable
+            });
+        }
+    }
```
Should we have a comment noting that all of these things are for implementing a particular interface?

reseul replied:
Sure. What would be the customary comment decoration?

---
In reply to: [159337071](https://github.com/Microsoft/reactxp/pull/404#discussion_r159337071) [](ancestors = 159337071)

rigdern replied:
I don't know if we have an established pattern. Something like `// From SomeThing interface`?

---
In reply to: [159342936](https://github.com/Microsoft/reactxp/pull/404#discussion_r159342936) [](ancestors = 159342936,159337071)

rigdern commented on src/windows/Button.tsx:
```diff
@@ -0,0 +1,211 @@
+
+    updateNativeTabIndex(): void {
+        if (this._focusableElement) {
+            let tabIndex: number | undefined = this.getTabIndex();
```
>: number | undefined [](start = 24, length = 20)

Is it preferred to allow TypeScript to infer the type in cases like this?

reseul replied:
I don't know, I've seen either way. Can be argued that it's clearer (to the viewer) this way.

---
In reply to: [159337210](https://github.com/Microsoft/reactxp/pull/404#discussion_r159337210) [](ancestors = 159337210)

rigdern commented on src/windows/Button.tsx:
```diff
@@ -0,0 +1,211 @@
+        // - true: keyboard focusable through any mean, receives keyboard input
+        // - false: not focusable at all, doesn't receive keyboard input
+        // The intermediate "focusable, but not in the tab order" case is not supported.
+        let windowsTabFocusable: boolean = !this.props.disabled && tabIndex !== undefined && tabIndex >= 0;
```
>!this.props.disabled && tabIndex !== undefined && tabIndex >= 0 [](start = 43, length = 63)

Should we put this in a helper? I see we have this logic a couple of times.

reseul replied:
may become unclear. There are subtle differences... sometimes we have a "disabled", sometime not, etc.

---
In reply to: [159337425](https://github.com/Microsoft/reactxp/pull/404#discussion_r159337425) [](ancestors = 159337425)

rigdern commented on src/windows/Button.tsx:
```diff
@@ -0,0 +1,211 @@
+                // ENTER triggers press on key down
+                if (key === KEY_CODE_ENTER) {
+                    this.props.onPress(keyEvent);
+                    return;
```
>return; [](start = 20, length = 7)

Why do we return here?

reseul replied:
leftover from times we had more code. Removing.

---
In reply to: [159337858](https://github.com/Microsoft/reactxp/pull/404#discussion_r159337858) [](ancestors = 159337858)

rigdern commented on src/windows/Button.tsx:
```diff
@@ -0,0 +1,211 @@
+                let key = keyEvent.keyCode;
+                // ENTER triggers press on key down
+                if (key === KEY_CODE_ENTER) {
+                    this.props.onPress(keyEvent);
```
>keyEvent [](start = 39, length = 8)

Is this the type of event this handler expects?

reseul replied:
This handler expects less (SyntheticEvent), but sine we already have the beefed up event, we pass that one.

---
In reply to: [159338363](https://github.com/Microsoft/reactxp/pull/404#discussion_r159338363) [](ancestors = 159338363)

rigdern commented on src/windows/Button.tsx:
```diff
@@ -0,0 +1,211 @@
+        let keyEvent = EventHelpers.toKeyboardEvent(e);
+        if (keyEvent.keyCode === KEY_CODE_SPACE) {
+            if (!this.props.disabled && this.props.onPress) {
+                this.props.onPress(keyEvent);
```
>keyEvent [](start = 35, length = 8)

Is this the type of event this handler expects?

reseul replied:
IT expects less (see similar comment elseswhere)

---
In reply to: [159338483](https://github.com/Microsoft/reactxp/pull/404#discussion_r159338483) [](ancestors = 159338483)

rigdern commented on src/native-common/FrontLayerViewManager.tsx:
```diff
@@ -134,6 +134,14 @@ export class FrontLayerViewManager {
         return null;
     }
 
+    public getActivePopupId() : string | null {
```
>getActivePopupId [](start = 11, length = 16)

Why wasn't this needed in the web implementation?

reseul replied:
Web is structured differently. FrontLayerViewManager calls RenderDOM.render() by passing a RootView element with an onDismissPopup prop.
In native-common it's RootView that calls into FrontLayerViewManager 

---
In reply to: [159348484](https://github.com/Microsoft/reactxp/pull/404#discussion_r159348484) [](ancestors = 159348484)

rigdern commented on src/native-common/Image.tsx:
```diff
@@ -96,8 +106,21 @@ export class Image extends React.Component<Types.ImageProps, {}> {
+    }
+
+    getChildContext() {
+        // Let descendant Types components know that their nearest Types ancestor is not an Types.Text.
```
>Types [](start = 26, length = 5)

I think all occurrences of `Types` here should be `RX`

reseul replied:
ok

---
In reply to: [159348619](https://github.com/Microsoft/reactxp/pull/404#discussion_r159348619) [](ancestors = 159348619)

rigdern commented on src/native-common/Text.tsx:
```diff
@@ -47,6 +60,17 @@ export class Text extends React.Component<Types.TextProps, {}> {
+    }
+
+    getChildContext() {
+        // Let descendant Types components know that their nearest Types ancestor is an Types.Text.
```
>Types [](start = 26, length = 5)

I think all occurrences of `Types` here should be `RX`

reseul replied:
ok

---
In reply to: [159348943](https://github.com/Microsoft/reactxp/pull/404#discussion_r159348943) [](ancestors = 159348943)

rigdern commented on src/native-common/TextInput.tsx:
```diff
@@ -117,7 +132,7 @@ export class TextInput extends React.Component<Types.TextInputProps, TextInputSt
     }
 
-    private _onSelectionChange = (selEvent: React.SyntheticEvent<TextInput>) => {
+    private _onSelectionChange = (selEvent: React.SyntheticEvent<any>) => {
```
>any [](start = 65, length = 3)

Why is this being changed?

reseul replied:
For consistency with other places, and because it hurt my eyes... This is a handler called by an RN.TextInput, so the correct type would have to be React.SyntheticEvent<RN.TextInput>, I think.

---
In reply to: [159349506](https://github.com/Microsoft/reactxp/pull/404#discussion_r159349506) [](ancestors = 159349506)

rigdern commented on src/native-desktop/RootView.tsx:
```diff
@@ -0,0 +1,165 @@
+        render() {
+            let content = super.render();
+
+            // Using "any" since onKeyDown/onKeyDown are not defined at RN.View property level
```
> onKeyDown/onKeyDown [](start = 32, length = 20)

This says `onKeyDown` twice

reseul replied:
lol

---
In reply to: [159357269](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357269) [](ancestors = 159357269)

rigdern commented on src/native-desktop/RootView.tsx:
```diff
@@ -0,0 +1,165 @@
+            this._focusManager = new FocusManager(undefined);
+        }
+
+        _onTouchStartCapture = (e: SyntheticEvent) => {
```
>_onTouchStartCapture [](start = 8, length = 20)

Does this get called for mouse input? If not, who calls `_updateKeyboardNavigationState` for mouse input?

reseul replied:
Yes. It's a "mouse down", basically.
One thing that I know RNW doesn't have it fully is the direct "on mouse move" event. It does have the "touch move", but that's something different (more like "mouse dragging"). I'll open an issue when I have a chance (and fix).

---
In reply to: [159357430](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357430) [](ancestors = 159357430)

rigdern commented on src/native-desktop/RootView.tsx:
```diff
@@ -0,0 +1,165 @@
+                this._isNavigatingWithKeyboard = isNavigatingWithKeyboard;
+
+                UserInterface.keyboardNavigationEvent.fire(isNavigatingWithKeyboard);
+            }
```
>} [](start = 12, length = 1)

How does the focus outline get rendered? It looks like the web version supports different focus outlines depending on whether the navigation was done with keyboard.

reseul replied:
We just defer to the UWP system focus rectangles (black rectangles) for now. Its displaying basically mimics the flag we have for "keyboard navigation on"", with some exceptions (the flag gets reset when pressing ESC, whereas I haven't found a way to convey this to UWP as well).
It's true web implementation offers 2 outlines. Also S4L disables that completely and uses an internal "focus outline" extension that shows a red rectangle. I chose not to cover that for now until it's clear why we have all these random initiatives and what we actually want.

---
In reply to: [159357484](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357484) [](ancestors = 159357484)

rigdern commented on src/typings/react-native-windows.d.ts:
```diff
@@ -0,0 +1,36 @@
+    interface FocusableProps extends RN.ViewProps {
+        isTabStop?                      : boolean;
+        tabIndex?                       : number;
+        tabNavigation?                  : string; // enum( 'local', 'cycle', 'once' );
```
>string; // enum( 'local', 'cycle', 'once' ); [](start = 42, length = 44)

I think you can use string types here:

`tabNavigation?: 'local' | 'cycle' | 'once'

reseul replied:
Will do. This trick doesn't seem to be used elsewhere, unfortunately.

---
In reply to: [159357725](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357725) [](ancestors = 159357725)

rigdern commented on src/typings/react-native-windows.d.ts:
```diff
@@ -0,0 +1,36 @@
+    // Focusable view related declarations
+    // ----------------------------------------------------------------------
+    interface FocusableProps extends RN.ViewProps {
+        isTabStop?                      : boolean;
```
>isTabStop?                      : boolean; [](start = 8, length = 42)

I think we prefer to avoid the column-like alignment because column-like alignment forces us to update the spacing of all the lines when we happen to add a new prop that has a name which is too long.

reseul replied:
react-native.d.ts is wildly inconsistent from your p.o.v. 
I'll follow your advice, hope I won't be getting the opposite as well...

---
In reply to: [159357801](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357801) [](ancestors = 159357801)

rigdern commented on src/typings/react-native.d.ts:
```diff
@@ -310,6 +317,11 @@ declare module 'react-native' {
+        // Windows only
+        onKeyDown? : Function;
+        onKeyUp? : Function;
+        tabNavigation? : string; // enum( 'local', 'cycle', 'once' );
```
>string [](start = 25, length = 6)

I think you can use string types here:

tabNavigation?: 'local' | 'cycle' | 'once'

reseul replied:
ok

---
In reply to: [159357972](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357972) [](ancestors = 159357972)

rigdern commented on src/windows/Animated.tsx:
```diff
@@ -0,0 +1,101 @@
+
+    focus() {
+        if (this._animatedComponent && this._animatedComponent._component) {
+            this._animatedComponent._component.focus();
```
>_component [](start = 36, length = 10)

What is `_component`? Why don't we need this indirection when implementing `setNativeProps`?

reseul replied:
_animatedComponent is a helper component used by the Animated package (you can check in react-native/Libraries/Animated/src/AnimatedImplementation.js, createAnimatedComponent).
It renders whatever you passed in (an RN.View in this case), with a ref function setting _component. So _component is the RN.View instance in this case.
Now, animatedComponent has a straight forwarder of setNativeProps to _component.setNativeProps, so no special indirection is needed there.
The "focus" method is outside this story.


---
In reply to: [159358021](https://github.com/Microsoft/reactxp/pull/404#discussion_r159358021) [](ancestors = 159358021)

rigdern commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,150 @@
+            if (key === KEY_CODE_ENTER) {
+                // Defer to base class
+                this._onPress(keyEvent);
+                return;
```
>return [](start = 16, length = 6)

is this return needed?

reseul replied:
nope

---
In reply to: [159358380](https://github.com/Microsoft/reactxp/pull/404#discussion_r159358380) [](ancestors = 159358380)

rigdern commented on src/windows/ReactXP.ts:
```diff
@@ -66,6 +66,7 @@ AccessibilityUtil.setAccessibilityPlatformUtil(AccessibilityPlatformUtil);
 // -- STRANGE THINGS GOING ON HERE --
 // See web/ReactXP.tsx for more details.
 
+// TODO change
```
What is this TODO for?

reseul replied:
To... remove itself!

---
In reply to: [159358539](https://github.com/Microsoft/reactxp/pull/404#discussion_r159358539) [](ancestors = 159358539)

rigdern commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,150 @@
+
+        let content: JSX.Element = super.render();
+
+        // The "in text parent" case requires a special nyi control.
```
>// The "in text parent" case requires a special nyi control. [](start = 8, length = 60)

Is this something we plan on doing later? If so, I think we should create a VSTS task for it and prefix this comment with `TODO` and reference the VSTS task number.

Also, I think it would be clearer if you created the branch for this case and then put the comment in that otherwise empty branch.

reseul replied:
Yes, it's the Xaml Link we talked about.
I think the task exists, yet adding the # here would be not very relevant to outside Msft, that's why I avoided it.  

I don't get the branch comment.

---
In reply to: [159358659](https://github.com/Microsoft/reactxp/pull/404#discussion_r159358659) [](ancestors = 159358659)

rigdern replied:
What I meant is to write something like this:

```
if (this.context && !this.context.isRxParentAText) {
  /* a bunch of implementation details that you have */
} else if (this.context && this.context.isRxParentAText) {
  // TODO: The "in text parent" case requires a React Native view that maps to XAML Hyperlink but this RN view isn't implemented yet.
}
```

---
In reply to: [159486509](https://github.com/Microsoft/reactxp/pull/404#discussion_r159486509) [](ancestors = 159486509,159358659)

rigdern commented on src/windows/ScrollView.tsx:
```diff
@@ -0,0 +1,51 @@
+
+        return (
+
+            <RN.ScrollView
```
extra new line

reseul replied:
ok

---
In reply to: [159358750](https://github.com/Microsoft/reactxp/pull/404#discussion_r159358750) [](ancestors = 159358750)

rigdern commented on src/windows/ScrollView.tsx:
```diff
@@ -0,0 +1,51 @@
+        var onKeyDownCallback = this.props.onKeyPress ?
+                // We have a callback function, call the wrapper
+                this._onKeyDown :
+                undefined;
```
This looks indented a little too much. I think it should be indented 4 spaces relative to the previous line.

reseul replied:
Ok (wth are the linters good for??)

---
In reply to: [159358840](https://github.com/Microsoft/reactxp/pull/404#discussion_r159358840) [](ancestors = 159358840)

rigdern commented on src/windows/TextInput.tsx:
```diff
@@ -0,0 +1,62 @@
+            tabIndex = {this.getTabIndex()}
+            onFocus = { (e: React.FocusEvent<any>) => this._onFocusEx(e, props.onFocus)}
+        />)
+            ;
```
I think the formatting is a little messed up. This stuff should be indented about 4 spaces more and the semicolon shouldn't be on a line by itselff.

reseul replied:
ok

---
In reply to: [159358874](https://github.com/Microsoft/reactxp/pull/404#discussion_r159358874) [](ancestors = 159358874)

rigdern commented on src/windows/TextInput.tsx:
```diff
@@ -0,0 +1,62 @@
+        return (
+            <RN.TextInput
+            {...props}
+            tabIndex = {this.getTabIndex()}
```
> =  [](start = 20, length = 3)

no spaces around the `=`

reseul replied:
ok

---
In reply to: [159358977](https://github.com/Microsoft/reactxp/pull/404#discussion_r159358977) [](ancestors = 159358977)

rigdern commented on src/windows/TextInput.tsx:
```diff
@@ -0,0 +1,62 @@
+            <RN.TextInput
+            {...props}
+            tabIndex = {this.getTabIndex()}
+            onFocus = { (e: React.FocusEvent<any>) => this._onFocusEx(e, props.onFocus)}
```
>onFocus [](start = 12, length = 7)

Are you stomping on the super class's `onFocus` handler?

reseul replied:
I think I was trying to be fancy by not having assumptions on what the base class passed in the props. The arrow function though is created one each render now, kind of bad. I'll try to reshuffle.

---
In reply to: [159359055](https://github.com/Microsoft/reactxp/pull/404#discussion_r159359055) [](ancestors = 159359055)

rigdern commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,371 @@
+import { applyFocusableComponentMixin, FocusManagerFocusableComponent, FocusManager } from '../native-desktop/utils/FocusManager';
+
+const KEY_CODE_ENTER = 13;
+const KEY_CODE_SPACE = 32;
```
Should we define these somewhere common rather than repeating them across multiple files?

reseul replied:
Maybe, but not sure where.

---
In reply to: [159359132](https://github.com/Microsoft/reactxp/pull/404#discussion_r159359132) [](ancestors = 159359132)

rigdern commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,371 @@
+
+        if (!!this.props.restrictFocusWithin !== !!nextProps.restrictFocusWithin) {
+            console.error('View: restrictFocusWithin is readonly and changing it during the component life cycle has no effect');
+        } else if (!!this.props.limitFocusWithin !== !!nextProps.limitFocusWithin) {
```
>limitFocusWithin [](start = 32, length = 16)

Are `restrictFocusWithin` and `limitFocusWithin` mutually exclusive? If not, we should use a second `if` rather than an `else if`

reseul replied:
True, copied from elsewhere. I'll change both places.

---
In reply to: [159359222](https://github.com/Microsoft/reactxp/pull/404#discussion_r159359222) [](ancestors = 159359222)

rigdern commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,371 @@
+            this._focusManager.release();
+        }
+        // Mixins may have added this
+        if (super.componentWillUnmount !== undefined) {
```
>if (super.componentWillUnmount !== undefined) { [](start = 8, length = 47)

Can this be simplified to?:

`if (super.componentWillUnmount) {`

reseul replied:
Yeah, maybe. I kind of hate the "if (x.method)" syntax, but if that's the customary pattern...

---
In reply to: [159359262](https://github.com/Microsoft/reactxp/pull/404#discussion_r159359262) [](ancestors = 159359262)

rigdern commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,371 @@
+
+            // Define the handler for "onKeyDown" on first use, it's the safest way when functions
+            // called from super constructors are involved.
+            if (this._onKeyDown === undefined) {
```
>if (this._onKeyDown === undefined) { [](start = 12, length = 36)

Can this be simplified to?:

if (this._oknKeyDown) {

rigdern commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    getChildContext() {
+        // Let descendant Types components know that their nearest Types ancestor is not an Types.Text.
```
>Types [](start = 26, length = 5)

`Types` should be `RX`

reseul replied:
ok

---
In reply to: [159359502](https://github.com/Microsoft/reactxp/pull/404#discussion_r159359502) [](ancestors = 159359502)

rigdern commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,371 @@
+            // ENTER triggers press on key down
+            if (key === KEY_CODE_ENTER) {
+                this.props.onPress(keyEvent);
+                return;
```
>return [](start = 16, length = 6)

extra return

reseul replied:
ok

---
In reply to: [159359635](https://github.com/Microsoft/reactxp/pull/404#discussion_r159359635) [](ancestors = 159359635)

rigdern commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,149 @@
+
+    getTabIndex(): number | undefined {
+        // Focus Manager may override this
+        return 0;
```
Why don't we reference `this.props.tabIndex`? It looks like it's a part of `Types.LinkProps`.

reseul replied:
Aha, The 0.42 RXP didn't have it. 
The current one does. Weirdly enough this tabIndex was added to LinkProps instead of making the LinkProps derive from that CommonAccessibilityProps or so..

---
In reply to: [159584978](https://github.com/Microsoft/reactxp/pull/404#discussion_r159584978) [](ancestors = 159584978)

rigdern commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,367 @@
+    }
+}
+
+// A value for tabIndex (even <0) marks a View as being potentially keyboard focusable
```
Why is it focusable with a `tabIndex` of `-1`? That seems different than the web implementation.

What does "keyboard focusable" mean? That you can hit the tab key to focus the View?

reseul replied:
"keyboard focusable" is meant to be used when differentiating from "mouse focusable".
Otherwise, we use "tab focusable" for being able to keyboard focus by tabbing.

Now, I decided to keep the few views with tabIndex==-1 under the FocusManager supervision in order to be able to implement the "focus the tabIndex==-1 component" trick. I'll take a new look, with all the changes you guys are asking for.

---
In reply to: [159586940](https://github.com/Microsoft/reactxp/pull/404#discussion_r159586940) [](ancestors = 159586940)

rigdern commented on src/native-desktop/UserInterface.tsx:
```diff
@@ -0,0 +1,29 @@
+        this.keyboardNavigationEvent.subscribe(this._keyboardNavigationStateChanged);
+    }
+
+    isNavigatingWithKeyboard(): boolean {
```
You put this under `native-desktop` but I wonder if this would be useful in non-desktop scenarios (e.g. an iPad with a physical keyboard). I believe Howard has been working on some iPad stuff. You might want to loop him in.

reseul commented:
Have to make a fix to align with PR feedback changes in RN for W, then I'll go through the latest comments.

erictraut commented on src/native-desktop/utils/FocusManager.ts:
```diff
@@ -0,0 +1,247 @@
+    });
+
+    // Implement 'setTabIndexOverride'
+    Component.prototype['setTabIndexOverride'] = function (this: FocusManagerFocusableComponentInternal, tabIndex: number) {
```
I agree with Adam's comment here.

reseul replied:
Yeah, I can do that.


---
In reply to: [159583189](https://github.com/Microsoft/reactxp/pull/404#discussion_r159583189) [](ancestors = 159583189)

reseul replied:
Same answer.

---
In reply to: [159785346](https://github.com/Microsoft/reactxp/pull/404#discussion_r159785346) [](ancestors = 159785346)

erictraut commented on src/typings/react-native.d.ts:
```diff
@@ -267,6 +267,13 @@ declare module 'react-native' {
         shouldRasterizeIOS? : boolean;
+
+        // Windows
+        onKeyDown?: Function;
```
Yikes, I don't like the idea of adding these props to all components. These should be added selectively only to the components where they make sense. Some of these are even duplicates of existing props, for example the onHoverStart/End for RX.Button.

Any place that we're adding props for keyboard or mouse we should be asking ourselves why we don't also add support for these same methods on the web (and eventually MacOS). They should be equally applicable.

Also, any callbacks should be strongly typed, not generic "Function" types.

reseul replied:
You mentioned RX.Button, so you may think this change is in Typings.ts. It's actually in react-native.d.ts.
onKeyDown handles bubbled events raised by children.
onMouse* are supported by a range of RN for W views. They map to RX.ViewPropShared.onMouse* properties, similar to what the web counterpart does with <div>.onMouse* events.

---
In reply to: [159786010](https://github.com/Microsoft/reactxp/pull/404#discussion_r159786010) [](ancestors = 159786010)

erictraut replied:
Ah, my bad. Yes, for some reason, I thought this was Typings.ts. Never mind.

erictraut commented on src/typings/react-native.d.ts:
```diff
@@ -310,6 +317,11 @@ declare module 'react-native' {
+        onKeyDown? : Function;
+        onKeyUp? : Function;
+        tabNavigation? : 'local' | 'cycle' | 'once';
+        disableKeyboardBasedScrolling?: boolean;
```
disableKeyboardBasedScrolling needs to be documented. Is it windows-specific? If so, it should also have a  comment.

reseul replied:
This is RN for Windows only, but we don't expose it over RX API.  I meant for the ""windows only" comment to apply to the whole paragraph, including disableKeyboardBasedScrolling. Where should I document more? Note this is react-native.d.ts, not Typings.ts


---
In reply to: [159786086](https://github.com/Microsoft/reactxp/pull/404#discussion_r159786086) [](ancestors = 159786086)

erictraut commented on src/typings/react-native.d.ts:
```diff
@@ -310,6 +317,11 @@ declare module 'react-native' {
         // iOS
         scrollIndicatorInsets?: {top: number, left: number, bottom: number, right: number };
+        // Windows only
+        onKeyDown? : Function;
```
Why do we need keydown/up methods in a scroll view? In the unlikely event that someone needs this, they can install the callbacks on a sub-view.

reseul replied:
RX.ScrollViewProps contains onKeyPress, (inherited from RX.ViewPropsShared, along with other props that don't really work in ScrollView...). onKeyPress is used at least in the VLV to implement app-side arrow based focus navigation.
The Web implementation passes onKeyPress to the backing <div> (just when using custom scrollbars, not sure why).
For Windows I decided to handle the onKeyDown bubbled React events from whatever "keyboard focusable" children of the scroll view.  

---
In reply to: [159786141](https://github.com/Microsoft/reactxp/pull/404#discussion_r159786141) [](ancestors = 159786141)

erictraut commented on src/typings/react-native.d.ts:
```diff
@@ -397,6 +409,8 @@ declare module 'react-native' {
         // macOS only property for submitting the text on enter
         submitTextOnEnter?: boolean;
+        // Windows only
+        tabIndex?: number;
```
Is tabIndex really needed for ScrollView? If so, why is it not applicable to web as well?

reseul replied:
This one is part of the RN.TextInputProps, and yes, it's the Windows flavor of RN that supports it.
RX.TextInput maps to <textarea> or <input> in the web implementation. RX.TextInputProps does contain tabIndex, which is not conveyed to the HTML elements (it should!), yet the web flavor of FocusManager does set the tabIndex DOM property on those elements liberally.
We needed something similar in RN for W. The resultant of RX.TextInputProps.tabIndex & FocusManager overrides is pushed to RN.TextInput through this property.

---
In reply to: [159786215](https://github.com/Microsoft/reactxp/pull/404#discussion_r159786215) [](ancestors = 159786215)

erictraut commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,149 @@
+
+import React = require('react');
+import RNW = require('react-native-windows');
+import {applyFocusableComponentMixin, FocusManagerFocusableComponent} from '../native-desktop/utils/FocusManager';
```
Curly braces should have spaces.

erictraut commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,149 @@
+import PropTypes = require('prop-types');
+
+import EventHelpers from '../native-common/utils/EventHelpers';
+import {Link as LinkCommon } from '../native-common/Link';
```
See above.

erictraut commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,149 @@
+const UP_KEYCODES = [KEY_CODE_SPACE];
+
+export interface LinkContext {
+    isRxParentAText?: boolean;
```
See question above about isRxParentAText. I don't understand why this is needed in native.

reseul replied:
Basically it covers a case that *could* be covered by React Native for Windows (either by mitigating, or by implementing the right approach), but it currently doesn't.

Our RX.Link is using a plain RN.Text for iOs/Android, whereas for Windows we encapsulate one of those in a RNW.FocusableWindows control to give it keyboard focusing capability.
Problem is that currently inlining such a construct in a parent RN.Text (when embedding a link in a longer text) crashes RNW completely.

I chose to mitigate this for now by avoiding the RNW.FocusableWindows encapsulation when the parent is an RN.Text generating RX.Text, with the hope that a NYI XamlLink based view would come to help this scenario, or maybe all of them (there are some tradeoffs between accessibility and select-ability, so we're not sure of the solution yet).

But if this tracking is not allowed on RXP side, what I'm left with is removing RX.Link from the story for now, and we'll see what future reserves us.




---
In reply to: [159786422](https://github.com/Microsoft/reactxp/pull/404#discussion_r159786422) [](ancestors = 159786422)

erictraut commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,149 @@
+                handledKeyUpKeys: UP_KEYCODES,
+                onKeyDown: this._onKeyDown,
+                onKeyUp: this._onKeyUp,
+                onFocus: this._onFocus,
```
Remove extraneous comma

erictraut commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,149 @@
+
+            content = (
+                <RNW.FocusableWindows
+                    {...focusableViewProps}
```
Curly braces should have spaces.

erictraut commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,149 @@
+    focus() {
+        if (this._focusableElement &&
+            this._focusableElement.focus) {
+                this._focusableElement.focus();
```
Please fix indenting for clarity.

erictraut commented on src/windows/Link.tsx:
```diff
@@ -0,0 +1,149 @@
+    blur() {
+        if (this._focusableElement &&
+            this._focusableElement.blur) {
+                this._focusableElement.blur();
```
Please fix indenting for clarity.

erictraut commented on src/windows/ScrollView.tsx:
```diff
@@ -0,0 +1,50 @@
+
+        return (
+            <RN.ScrollView
+                {...props}
```
Curly braces should have spaces.

erictraut commented on src/windows/ScrollView.tsx:
```diff
@@ -0,0 +1,50 @@
+                onKeyDown={ onKeyDownCallback }
+                keyboardShouldPersistTaps={ keyboardShouldPersistTaps }
+                tabNavigation={ this.props.tabNavigation }
+                disableKeyboardBasedScrolling={true}
```
See above.

erictraut commented on src/windows/TextInput.tsx:
```diff
@@ -0,0 +1,62 @@
+
+        return (
+            <RN.TextInput
+            {...props}
```
Please fix indentation and add spaces in curly braces.

erictraut commented on src/windows/TextInput.tsx:
```diff
@@ -0,0 +1,62 @@
+            let tabIndex: number | undefined = this.getTabIndex();
+            this._textInputRef.setNativeProps({
+                tabIndex: tabIndex,
+                value: this.state.inputValue, // mandatory for some reason
```
Remove extraneous comma.

erictraut commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,367 @@
+import Types = require('../common/Types');
+import PropTypes = require('prop-types');
+
+import {View as ViewCommon} from '../native-common/View';
```
Curly braces should have spaces.

erictraut commented on src/windows/View.tsx:
```diff
@@ -0,0 +1,367 @@
+
+            content = (
+                <RNW.FocusableWindows
+                    {...focusableViewProps}
```
Curly braces should have spaces.

reseul commented on src/native-desktop/utils/FocusManager.ts:
```diff
@@ -0,0 +1,247 @@
+                return action.call(this, origCallback, arguments);
+            };
+        } else {
+            console.error('FocusableComponentMixin: ' + methodName + ' error!');
```
The methods have to exist on components, they're mandatory. I can make the error message friendlier.

---
In reply to: [159582412](https://github.com/Microsoft/reactxp/pull/404#discussion_r159582412) [](ancestors = 159582412)

reseul commented on src/native-desktop/RootView.tsx:
```diff
@@ -0,0 +1,165 @@
+            focusManager: PropTypes.object
+        };
+
+        _focusManager: FocusManager;
```
I really wanted to mark these accordingly, but any attempt to do that triggers an error later in the file (where we generate the new decorated classes) that looks like this "Property '_focusManager' of exported class expression may not be private or protected.".

The 0.42 flavor of this was simpler, I had just one class extensions. The current RXP versions brought multiple RootView flavors (with store/with props), so I tried the mixin approach to "extend" both base RootView classes with one set of code.


---
In reply to: [159783697](https://github.com/Microsoft/reactxp/pull/404#discussion_r159783697) [](ancestors = 159783697)

reseul commented on src/native-desktop/utils/FocusManager.ts:
```diff
@@ -0,0 +1,247 @@
+        super(parent);
+    }
+
+    protected /* static */ addFocusListenerOnComponent(component: React.Component<any, any>, onFocus: () => void): void {
```
The signature is shared with the Web flavor of FocusManager, whereas FocusManagerFocusableComponentInternal  is specific to the "desktop" implementations only. 
Not sure how to proceed to make it look better and less nasty.

---
In reply to: [159784667](https://github.com/Microsoft/reactxp/pull/404#discussion_r159784667) [](ancestors = 159784667)

reseul commented on src/native-desktop/utils/FocusManager.ts:
```diff
@@ -0,0 +1,247 @@
+            // Defer the focusing to let the view finish its initialization and to allow for manual focus setting (if any)
+            // to be processed (the asynchronous nature of focus->onFocus path requires a delay)
+            FocusManager._resetFocusTimer = setTimeout(() => {
+                FocusManager._resetFocusTimer = undefined;
```
The handler touches just the FocusManager._currentFocusedComponent, which could be stale. The check below should also have a "|| storedComponent.removed". I'll fix here and wherever I see the pattern.
There won't be a crash without the fix, but the code may decide there's no need to call "focusFirst" when it actually is.

---
In reply to: [159784893](https://github.com/Microsoft/reactxp/pull/404#discussion_r159784893) [](ancestors = 159784893)

reseul commented on src/windows/Button.tsx:
```diff
@@ -0,0 +1,212 @@
+            >
+                <RN.Animated.View
+                    {...internalProps}
+                    onMouseEnter={this._onMouseEnter}
```
RX.ButtonProps has onHoverStart/etc., there's no conflict. I kept RNW.FocusableWindows solely for the keyboard related stuff.


---
In reply to: [159584124](https://github.com/Microsoft/reactxp/pull/404#discussion_r159584124) [](ancestors = 159584124)

reseul commented on src/native-desktop/RootView.tsx:
```diff
@@ -0,0 +1,165 @@
+                }
+
+                this._isNavigatingWithKeyboardUpateTimer = setTimeout(() => {
+                    this._isNavigatingWithKeyboardUpateTimer = undefined;
```
Not in this case, we don't dereference/call any component

---
In reply to: [159784381](https://github.com/Microsoft/reactxp/pull/404#discussion_r159784381) [](ancestors = 159784381)

erictraut commented:
@reseul, what's the current status of this PR? Do you have more changes you'd like to make before I merge it, or is it good to go?

reseul commented:
@erictraut, not ready yet. I have to fix that ugly casting + a little of the demixinification. 
I'll send a ping when done.

reseul commented:
@erictraut , ok, now this is ready in my opinion.
On full disclosure, I added these since fist PR was sent:
- changed the way we instantiate FocusableWindow: RNW has  "factory" createFocusableComponent(component) now.
- made the "calling focus on components with tabIndex -1" a little more accurate.
- I kept the IsRxTextParent hack in for now.

Now, you and @rigdern had a bunch of comments, and I *think* I address all of them, at least with some answers if not code changes, but I guess you should confirm.
As far as I'm concerned, this should go in! :)
